File: kiosk/src/lib/scanMatch.test.ts
import { expect, it } from 'vitest';

import {
  buildScanIndex, matchAssetOrSerial, matchScan, scanTypeFor, type ScanAsset,
} from './scanMatch';

const rows: ScanAsset[] = [
  {
    id: 'a-1', asset_id: '10042', name: 'Rack 4 switch', rfid: '000000000000100348',
    serial_number: 'sn-4242', make: 'Cisco', model: 'Nexus 9000', make_model: 'Cisco Nexus 9000',
  },
  {
    id: 'a-2', asset_id: '10043', name: 'Patch panel', rfid: null,
    serial_number: 'FDO2140X9ZZ', make: null, model: null, make_model: '',
  },
  {
    id: 'a-3', asset_id: 'A-99', name: 'No tags', rfid: '', serial_number: null, make_model: '',
  } as ScanAsset,
];

const index = buildScanIndex(rows);

it('matches an RFID tag with or without its leading zeros, any case', () => {
  for (const raw of ['000000000000100348', '100348', '  100348  ']) {
    expect(matchScan(index, raw)).toMatchObject({ kind: 'rfid', asset: { id: 'a-1' } });
  }
});

it('matches an RFID tag case-insensitively', () => {
  const hex = buildScanIndex([{ ...rows[0], rfid: '00000000E28011AB' }]);
  expect(matchScan(hex, 'e28011ab')).toMatchObject({ kind: 'rfid' });
  expect(matchScan(hex, 'E28011AB')).toMatchObject({ kind: 'rfid' });
});

it('matches an asset ID, case-insensitively', () => {
  expect(matchScan(index, '10043')).toMatchObject({ kind: 'asset_id', asset: { id: 'a-2' } });
  expect(matchScan(index, 'a-99')).toMatchObject({ kind: 'asset_id', asset: { id: 'a-3' } });
});

it('matches a serial number, case-insensitively and trimmed', () => {
  expect(matchScan(index, 'SN-4242')).toMatchObject({ kind: 'serial', asset: { id: 'a-1' } });
  expect(matchScan(index, ' fdo2140x9zz ')).toMatchObject({ kind: 'serial', asset: { id: 'a-2' } });
});

it('tries RFID before asset ID before serial', () => {
  // One row's asset_id is another row's serial; the asset_id wins.
  const clash = buildScanIndex([
    { ...rows[0], id: 'by-serial', serial_number: 'X1', asset_id: 'zzz', rfid: null },
    { ...rows[1], id: 'by-asset', asset_id: 'X1', serial_number: null, rfid: null },
    { ...rows[2], id: 'by-rfid', rfid: 'X1', asset_id: 'yyy', serial_number: null },
  ]);
  expect(matchScan(clash, 'X1')).toMatchObject({ kind: 'rfid', asset: { id: 'by-rfid' } });
});

it('returns null for an unknown value, a blank one, and empty/missing tags', () => {
  expect(matchScan(index, 'nope123')).toBeNull();
  expect(matchScan(index, '')).toBeNull();
  expect(matchScan(index, '   ')).toBeNull();
  // a-3 has rfid '' and no serial: neither may become a catch-all key
  expect(matchScan(index, '0')).toBeNull();
});

it('an empty roster matches nothing', () => {
  expect(matchScan(buildScanIndex([]), '100348')).toBeNull();
});

it('scanTypeFor reports rfid for a tag and barcode for everything else', () => {
  expect(scanTypeFor('rfid')).toBe('rfid');
  expect(scanTypeFor('asset_id')).toBe('barcode');
  expect(scanTypeFor('serial')).toBe('barcode');
});

it('matchAssetOrSerial matches an asset ID or a serial, and never an RFID tag', () => {
  expect(matchAssetOrSerial(index, '10043')).toMatchObject({ kind: 'asset_id', asset: { id: 'a-2' } });
  expect(matchAssetOrSerial(index, ' sn-4242 ')).toMatchObject({ kind: 'serial', asset: { id: 'a-1' } });
  // a-1's RFID tag, padded and trimmed: matchScan finds it, this must not
  expect(matchScan(index, '100348')).toMatchObject({ kind: 'rfid' });
  expect(matchAssetOrSerial(index, '100348')).toBeNull();
  expect(matchAssetOrSerial(index, '000000000000100348')).toBeNull();
});

it('matchAssetOrSerial tries the asset ID before the serial', () => {
  const clash = buildScanIndex([
    { ...rows[0], id: 'by-serial', serial_number: 'X1', asset_id: 'zzz', rfid: null },
    { ...rows[1], id: 'by-asset', asset_id: 'X1', serial_number: null, rfid: null },
  ]);
  expect(matchAssetOrSerial(clash, 'x1')).toMatchObject({ kind: 'asset_id', asset: { id: 'by-asset' } });
});

it('matchAssetOrSerial returns null for a blank or unknown value', () => {
  expect(matchAssetOrSerial(index, '')).toBeNull();
  expect(matchAssetOrSerial(index, '  ')).toBeNull();
  expect(matchAssetOrSerial(index, 'nope123')).toBeNull();
});

File: Android_Kiosk_App/app/src/test/java/com/serversherpa/kiosk/core/scan/ScanMatchTest.kt
package com.serversherpa.kiosk.core.scan

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

private data class A(
    override val id: String, override val assetId: String, override val name: String? = null,
    override val rfid: String? = null, override val serialNumber: String? = null, override val makeModel: String = "",
) : ScanAsset

class ScanMatchTest {
    private val rack = A("1", "A-100", "Rack", rfid = "000000000000000000100348", serialNumber = "SN-1")
    private val server = A("2", "A-200", "Server", rfid = null, serialNumber = "sn-2")
    private val index = buildScanIndex(listOf(rack, server))

    @Test fun rfidThenAssetIdThenSerial() {
        assertEquals(ScanMatchKind.RFID, matchScan(index, "100348")?.kind)
        assertEquals(ScanMatchKind.RFID, matchScan(index, "000000000000000000100348")?.kind)
        assertEquals("1", matchScan(index, "a-100")?.asset?.id)
        assertEquals(ScanMatchKind.ASSET_ID, matchScan(index, "a-100")?.kind)
        assertEquals(ScanMatchKind.SERIAL, matchScan(index, "SN-2")?.kind)
        assertNull(matchScan(index, "nothing"))
        assertNull(matchScan(index, "   "))
    }

    @Test fun firstRowWinsOnDuplicateKeys() {
        val dup = A("9", "A-100", "Other")
        val i = buildScanIndex(listOf(rack, dup))
        assertEquals("1", matchScan(i, "A-100")?.asset?.id)
        assertEquals(2, i.size)
    }

    @Test fun assetOrSerialNeverMatchesByTag() {
        assertNull(matchAssetOrSerial(index, "100348"))
        assertEquals(ScanMatchKind.ASSET_ID, matchAssetOrSerial(index, "A-100")?.kind)
        assertEquals(ScanMatchKind.SERIAL, matchAssetOrSerial(index, "sn-1")?.kind)
    }

    @Test fun scanTypeWire() {
        assertEquals("rfid", scanTypeFor(ScanMatchKind.RFID))
        assertEquals("barcode", scanTypeFor(ScanMatchKind.ASSET_ID))
        assertEquals("barcode", scanTypeFor(ScanMatchKind.SERIAL))
    }
}
